lib/very_good_infinite_list.dart

```diff
@@ -1,459 +1,274 @@
-    this.exception = _ListException.none,
-  });
 
-  final int currentIndex;
-  final List<T> items;
-  final _ListStatus status;
-  final bool hasReachedMax;
-  final _ListException exception;
-
-  _ListState<T> copyWith({
-    required _ListStatus status,
-    int? currentIndex,
-    List<T>? items,
-    bool? hasReachedMax,
-    _ListException? exception,
-  }) {
-    return _ListState<T>(
-      currentIndex: currentIndex ?? this.currentIndex,
-      items: items ?? this.items,
-      status: status,
-      hasReachedMax: hasReachedMax ?? this.hasReachedMax,
-      exception: exception ?? this.exception,
-    );
+  void _attemptFetch() {
+    if (_isAtEnd &&
+        !widget.hasReachedMax &&
+        !widget.isLoading &&
+        !widget.hasError) {
+      _debounce(widget.onFetchData);
+    }
   }
-}
-
-class _ListController<T> extends ValueNotifier<_ListState<T>> {
-  _ListController(this.itemLoader) : super(_ListState<T>());
 
-  final ItemLoader<T> itemLoader;
-
-  void fetch({int limit = 20}) async {
-    if (value.hasReachedMax) return;
-
-    if (value.currentIndex == 0 && value.items.isEmpty) {
-      value = value.copyWith(status: _ListStatus.loading);
+  bool get _isAtEnd {
+    if (widget.items.isEmpty) {
+      return true;
     }
 
-    if (value.status.isFailure) {
-      if (value.exception.value is InfiniteListException) {
-        value = value.copyWith(
-          exception: _ListException.none,
-          status: _ListStatus.loading,
-        );
-      } else {
-        value = value.copyWith(
-          exception: _ListException.none,
-          status: value.items.isEmpty ? _ListStatus.loading : value.status,
-        );
-      }
+    if (!_scrollController!.hasClients) {
+      return false;
     }
 
-    try {
-      final items = await itemLoader(limit, start: value.currentIndex);
+    final maxScroll = _scrollController!.position.maxScrollExtent;
+    final currentScroll = _scrollController!.offset;
+    return currentScroll >= (maxScroll - widget.scrollExtentThreshold);
+  }
 
-      if (items == null || items.isEmpty) {
-        value = value.copyWith(
-          hasReachedMax: true,
-          status: _ListStatus.success,
+  WidgetBuilder get _loadingBuilder =>
+      widget.loadingBuilder ??
+      (context) {
+        return const Center(
+          child: CircularProgressIndicator(),
         );
-        return;
-      }
-
-      value = _ListState(
-        currentIndex: value.currentIndex + items.length,
-        items: List.of(value.items)..addAll(items),
-        status: _ListStatus.success,
-      );
-    } catch (e) {
-      value = value.copyWith(
-        status: _ListStatus.failure,
-        exception: _ListException(e),
-      );
-    }
+      };
+
+  WidgetBuilder get _errorBuilder =>
+      widget.errorBuilder ??
+      (context) {
+        return const Center(
+          child: Text('Error'),
+        );
+      };
+
+  @override
+  Widget build(BuildContext context) {
+    return ListView(
```

Comment:
I would highly recommend using `ListView.builder` so that we get the performance benefits and aren't forced to maintain all of the items in memory even when they aren't visible.

Reply:
Done!